File: 03-poo/16-integracion-poo/scala/ejemplos/IntegracionCursos.scala
trait Evaluable:
  def progreso: Double
case class Participante(nombre: String)
class Inscripcion(val participante: Participante) extends Evaluable:
  private var avance = 0.0
  def actualizar(valor: Double): Unit = { require(valor >= 0 && valor <= 100); avance = valor }
  def progreso: Double = avance
class CursoIntegrado(val titulo: String):
  private var inscripciones = List.empty[Inscripcion]
  def inscribir(p: Participante): Unit = inscripciones ::= new Inscripcion(p)
  def total: Int = inscripciones.size

@main def integracionCursos(): Unit =
  val c = new CursoIntegrado("POO")
  c.inscribir(Participante("Aprendiz"))
  println(c.total)

File: 03-poo/16-integracion-poo/java/ejemplos/IntegracionCursos.java
import java.util.ArrayList;
import java.util.List;
public class IntegracionCursos {
    interface Evaluable { double progreso(); }
    static class Participante { final String nombre; Participante(String nombre){this.nombre=nombre;} }
    static class Inscripcion implements Evaluable { final Participante participante; private double avance; Inscripcion(Participante p){participante=p;} void actualizar(double avance){ if(avance<0||avance>100) throw new IllegalArgumentException(); this.avance=avance; } public double progreso(){return avance;} }
    static class Curso { final String titulo; final List<Inscripcion> inscripciones=new ArrayList<>(); Curso(String titulo){this.titulo=titulo;} void inscribir(Participante p){inscripciones.add(new Inscripcion(p));} }
    public static void main(String[] args){ Curso c=new Curso("POO"); c.inscribir(new Participante("Aprendiz")); System.out.println(c.inscripciones.size()); }
}
